pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,348 @@
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+func mergeVirtualServices(vServices []Config) (out []Config) {
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
+			delegatesMap[vs.Namespace+vs.Name] = vs
+			continue
+		}
+
+		// root vs
+		if isRootVs(rule) {
+			rootVses = append(rootVses, vs)
+			continue
+		}
+
+		// the others are normal vs without delegate
+		out = append(out, vs)
+	}
+
+	// 2. merge delegates and root
+	for _, root := range rootVses {
+		rootVs := root.Spec.(*networking.VirtualService)
+		notFound := false
+		mergedRoutes := []*networking.HTTPRoute{}
+		for _, route := range rootVs.Http {
+			// it is root vs with delegate
+			if route.Delegate != nil {
+				delegate, ok := delegatesMap[route.Delegate.Namespace+route.Delegate.Name]
+				if !ok {
+					log.Debugf("delegate virtual service %s/%s of %s/%s not found",
+						route.Delegate.Namespace, route.Delegate.Name, root.Namespace, root.Name)
+					// delegate not found, ignore the root as well
+					notFound = true
+					break
+				}
+				vs := delegate.Spec.(*networking.VirtualService)
+				merged := mergeHttpRoute(route, vs.Http)
+				mergedRoutes = append(mergedRoutes, merged...)
+			} else {
+				mergedRoutes = append(mergedRoutes, route)
+			}
+		}
+		if !notFound {
+			rootVs.Http = mergedRoutes
+			out = append(out, root)
+		}
+	}
+
+	return
+}
+
+// merge root's route with delegate's and the merged route number equals the delegate's.
+// if there is a conflict with root, the route is ignored
+func mergeHttpRoute(root *networking.HTTPRoute, delegate []*networking.HTTPRoute) []*networking.HTTPRoute {
+	root.Delegate = nil
+
+	out := make([]*networking.HTTPRoute, 0, len(delegate))
+	for _, subRoute := range delegate {
+		// suppose there are N1 match conditions in root, N2 match conditions in delegate
+		// if match condition of N2 is a subset of anyone in N1, this is a valid matching conditions
+		merged, conflict := mergeHTTPMatchRequests(root.Match, subRoute.Match)
+		if conflict {
+			log.Debugf("HTTPMatchRequests conflict: root root %s, delegate root %s", root.Name, subRoute.Name)
+			continue
+		}
+		subRoute.Match = merged
+
+		if subRoute.Name == "" {
+			subRoute.Name = root.Name
+		}
+		if subRoute.Rewrite == nil {
+			subRoute.Rewrite = root.Rewrite
+		}
+		if subRoute.Timeout == nil {
+			subRoute.Timeout = root.Timeout
+		}
+		if subRoute.Retries == nil {
+			subRoute.Retries = root.Retries
+		}
+		if subRoute.Fault == nil {
+			subRoute.Fault = root.Fault
+		}
+		if subRoute.Mirror == nil {
+			subRoute.Mirror = root.Mirror
+		}
+		if subRoute.MirrorPercentage == nil {
+			subRoute.MirrorPercentage = root.MirrorPercentage
+		}
+		if subRoute.CorsPolicy == nil {
```

Comment:
any way we can ensure new fields are added here?

Reply:
If we can introduce fuzz test, then it can be used to intercept new fields unset here.

Reply:
Can we create a test and assert the current field count if possible using reflection? If a new field is added, that test will fail and we can fix it?

Reply:
added fuzz test, could help in some degree